Consolidate and formalize conformance language (includes PR 23)

louberger opened the pull request:
also includes/builds on  #23

louberger committed:
Refactor scenarios section to make MPLS scenario the main and
	 IP and TSN supporting -- May decide to drop / move these.

louberger committed:
Revised section 2 - cover only MPLS layer and below

louberger committed:
fix from andy

louberger committed:
Minor edits: including from Andy

louberger committed:
fix capitalization of titles (consistent cammel case)

louberger committed:
Consolidate and formalize conformance language

louberger committed:
Add PEOF requirements sections

louberger committed:
travis/Makefile: add make travis target so formatted text and diff are captured in log

vb-balazs commented on draft-ietf-detnet-dp-sol-mpls-02.xml:
```diff
@@ -149,22 +147,33 @@
+  <t>
+    The intent is for this document to support different traffic types
+    being mapped over DetNet MPLS, but this is out side the scope of this
+    document. An example of such can be found in <xref
```
What do we ntend to say here? Does it refer to L2/L3 App-flow? Or somewhat else? What is covered in the IP data-plane draft?

vb-balazs commented on draft-ietf-detnet-dp-sol-mpls-02.xml:
```diff
@@ -453,19 +414,150 @@ System  |    +--------+       +--------+       +--------+   | System
-<figure align="center" anchor="fig_pw_detnet2"
- title="MPLS-Based DetNet (non-MPLS End System)">
+    <t>
+      [Authors note: TBD - should the rest of this section be cut?]
```
No, these are good and straight-fort  description for scenarios where "MPLS to the desktop" is not possible.

vb-balazs commented on draft-ietf-detnet-dp-sol-mpls-02.xml:
```diff
@@ -484,51 +576,107 @@ System  |    +--------+       +--------+       +--------+   | System
-  end-to-end.
- </t>
+  </section>
+  <section title="IEEE 802.1 TSN Over MPLS DetNet Data Plane Scenario"
```
This is a more general case. It can be any L2 not only TSN. For example if the IP End-systems are in the same subnet and connected via the MPLS domain. Worth to add a note?

vb-balazs commented on draft-ietf-detnet-dp-sol-mpls-02.xml:
```diff
@@ -484,51 +576,107 @@ System  |    +--------+       +--------+       +--------+   | System
+|   Appl.  |<-:Svc Proxy:--End to End Svc.--:Svc Proxy:->|   Appl.  |
++----------+  +---------+                   +---------+  +----------+
+|    TSN   |  |TSN| |Svc|<-- DetNet flow -->: Service :->|    TSN   |
++----------+  +---+ +---+    +----------+   +---------+  +----------+
```
Fix the right side Edge node. Service --> Svc + TSN. Same as on the left side.

vb-balazs commented on draft-ietf-detnet-dp-sol-mpls-02.xml:
```diff
@@ -484,51 +576,107 @@ System  |    +--------+       +--------+       +--------+   | System
+   stations operating over a TSN aware DetNet service running over an
+   MPLS network.  DetNet Edge Nodes sit at the boundary of a DetNet
+   domain. They are responsible for mapping non-DetNet aware traffic to
+   DetNet services.  They also support the imposition and disposition of
```
Add Layer-2:
"They are responsible for mapping non-DetNet aware **L2** traffic to"

vb-balazs commented on draft-ietf-detnet-dp-sol-mpls-02.xml:
```diff
@@ -1113,37 +1004,337 @@ Examples:
+   When ELs are used, the same EL value SHOULD be used for
+   all of the packets sent using a specific S-Label to force the flow to follow the
+   same path.   However, as previously stated in <xref target="dn-gen-encap-solution"/>, the use of ECMP in a DetNet domain is NOT RECOMMENDED.
+   </t>
```
Not recommended for DetNet flows, but may be used for non-DetNet traffic.

vb-balazs commented on draft-ietf-detnet-dp-sol-mpls-02.xml:
```diff
@@ -1113,37 +1004,337 @@ Examples:
+       and the controller plane, see <xref target="cp_considerations"/>.
+     </t>
+     <t>
+       After an app flow is identified for a received DetNet MPLS
```
Does it mean "elimination first"? Does this text set the order of PREOF functions?

vb-balazs commented on draft-ietf-detnet-dp-sol-mpls-02.xml:
```diff
@@ -1113,37 +1004,337 @@ Examples:
+     <t>
+       Note that an implementation MAY wish to constrain the maximum
+       number sequence numbers that are tracked, on platform-wide or per
+       flow basis.  Some implementations MAY support the provisioning of
```
What it means per flow basis? There is always a single seq.num per flow.

vb-balazs commented on draft-ietf-detnet-dp-sol-mpls-02.xml:
```diff
@@ -1113,37 +1004,337 @@ Examples:
+     <t>
+       A function that is related to PEF is the Packet Ordering Function
+       (POF).  In addition to PEF, implementations MAY support POF. When
+       supported, use of the POF for a particular app flow MUST be
```
POF can be also a stand alone function. This text binds it with PEF ("In addition to PEF ...").

vb-balazs commented on draft-ietf-detnet-dp-sol-mpls-02.xml:
```diff
@@ -1113,37 +1004,337 @@ Examples:
+     </t>
+     <section anchor="f-labels-ssl"
+              title="Service Sub-Layer Processing Node Processing">
+       
```
Does it mean "Node Processing of Service Sub-Layer Processing" ... ??

vb-balazs commented on draft-ietf-detnet-dp-sol-mpls-02.xml:
```diff
@@ -1113,37 +1004,337 @@ Examples:
+       by an implementation for outgoing DetNet flows. When supported,
+       the same app-flow data will be sent over multiple outgoing
+       forwarding sub-layer LSPs.  To support PRF an implementation MUST
+       support the setting of different sets of F-Labels.  Hereto, to
```
Compound flows may have different S-Label. So, add:
"different sets of F-Labels **and S-labels**"

louberger commented:
I'm confused by this. I thought an app flow always mapped to a single s 
label and that s label may map to multiple member flows using different f 
labels.  Are you saying a single app flow may use multiple s labels? If so, 
under what conditions?


----------
On March 4, 2019 8:27:27 AM Varga Balázs <notifications@github.com> wrote:

> vb-balazs commented on this pull request.
>
>
>
>> +       </t>
> +       <t>
> +         When sending a DetNet flow, Zero or more F-Labels MAY be added
> +         on top of an S-Label by the node pushing an S-Label. The
> +         F-Labels to be pushed when sending a particular app-flow MUST
> +         be provisioned per app-flow via configuration and the
> +         controller plane, see <xref target="cp_considerations"/>.  To
> +         allow for the omission of F-Labels, an implementation SHOULD
> +         also allow an outgoing interface to be provisioned.
> +     </t>
> +     <t>
> +       The Packet Replication Function (PRF) function MAY be supported
> +       by an implementation for outgoing DetNet flows. When supported,
> +       the same app-flow data will be sent over multiple outgoing
> +       forwarding sub-layer LSPs.  To support PRF an implementation MUST
> +       support the setting of different sets of F-Labels.  Hereto, to
>
> Compound flows may have different S-Label. So, add:
> "different sets of F-Labels **and S-labels**"
>
> -- 
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/detnet-wg/draft-ietf-detnet-dp-sol-mpls/pull/24#pullrequestreview-210136542

louberger commented:
Sorry I think this is a bad cut and paste oh, probably node processing 
should just be dropped.

----------
On March 4, 2019 8:27:22 AM Varga Balázs <notifications@github.com> wrote:

> vb-balazs commented on this pull request.
>
>
>
>> +       platform-wide or per flow basis.  Some implementations MAY
> +       support the provisioning of this number on either a platform-wide
> +       or per flow basis.  The number of out of order packets that can
> +       be processed also impacts the latency of a flow.
> +     </t>
> +   </section>
> +  </section>
> +  <section anchor="f-labels" title="F-Labels">
> +     <t>
> +       F-Labels are support the DetNet forwarding sub-layer. F-Labels
> +       are used to provide LSP-based connectivity between DetNet service
> +       sub-layer processing nodes.
> +     </t>
> +     <section anchor="f-labels-ssl"
> +              title="Service Sub-Layer Processing Node Processing">
> +
>
> Does it mean "Node Processing of Service Sub-Layer Processing" ... ??
>
> -- 
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/detnet-wg/draft-ietf-detnet-dp-sol-mpls/pull/24#pullrequestreview-210136512

louberger commented:
Sure....


----------
On March 4, 2019 8:26:37 AM Varga Balázs <notifications@github.com> wrote:

> vb-balazs commented on this pull request.
>
>
>
>> ++------.---+  +--.+ +-.-+    +---.----.-+   +--.+ +-.-+  +----.-----+
> +       :   Link  :    /  ,-----.  \   :  Link  :   /  ,-----.  \
> +       +.........+    +-[  Sub  ]-+   +........+   +-[  Sub  ]-+
> +                        [Network]                    [Network]
> +                         `-----'                      `-----'
> +
> +        |<- TSN ->| |<------ DetNet MPLS ------>| |<-- TSN --->|
> +]]></artwork>
> +</figure>
> +
> +  <t>
> +   <xref target="fig_tsn_mpls_detnet"/> shows IEEE 802.1 TSN end
> +   stations operating over a TSN aware DetNet service running over an
> +   MPLS network.  DetNet Edge Nodes sit at the boundary of a DetNet
> +   domain. They are responsible for mapping non-DetNet aware traffic to
> +   DetNet services.  They also support the imposition and disposition of
>
> Add Layer-2:
> "They are responsible for mapping non-DetNet aware **L2** traffic to"
>
> -- 
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/detnet-wg/draft-ietf-detnet-dp-sol-mpls/pull/24#pullrequestreview-210136200

louberger commented:
Agreed. But the maximum number of sequence numbers that are tracked may be 
per platform or per flow.


----------
On March 4, 2019 8:27:13 AM Varga Balázs <notifications@github.com> wrote:

> vb-balazs commented on this pull request.
>
>
>
>> +       After an app flow is identified for a received DetNet MPLS
> +       packet, as described above, an implementation MUST check if PEF
> +       is configured for that app flow.  When configured the
> +       implementation MUST track the sequence number contained in
> +       received d-CWs and MUST ensure that duplicate (replicated)
> +       instances of a particular sequence number are discarded.  The
> +       specific mechanisms used for an implementation to identify which
> +       received packets are duplicates and which are new is an
> +       implementation choice.  Note that per <xref target="dn-sn"/>
> +       the sequence number field length may be 16 or 28 bits, and the
> +       field value can wrap.
> +     </t>
> +     <t>
> +       Note that an implementation MAY wish to constrain the maximum
> +       number sequence numbers that are tracked, on platform-wide or per
> +       flow basis.  Some implementations MAY support the provisioning of
>
> What it means per flow basis? There is always a single seq.num per flow.
>
> -- 
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/detnet-wg/draft-ietf-detnet-dp-sol-mpls/pull/24#pullrequestreview-210136457

louberger commented:
Thanks! Will do.


----------
On March 4, 2019 8:26:30 AM Varga Balázs <notifications@github.com> wrote:

> vb-balazs commented on this pull request.
>
>
>
>> -|   |-------|._X_....|..DF1..|.__ ___.|..DF3..|...._X_.|-------|   |
> -|CE1|   |   |    \   |       |   X    |       |   /    |   |   |CE2|
> -|   |   |   |     \_.|..DF2..|._/ \__.|..DF4..|._/     |   |   |   |
> +
> +<figure anchor="fig_tsn_mpls_detnet" align="center"
> +title="A TSN over DetNet MPLS Enabled Network">
> +<artwork align="center"><![CDATA[
> + TSN             Edge          Transit        Edge        TSN
> + End System      Node           Node          Node        End System
> +                (T-PE)         (LSR)          (T-PE)
> +
> ++----------+  +.........+                   +.........+  +----------+
> +|   Appl.  |<-:Svc Proxy:--End to End Svc.--:Svc Proxy:->|   Appl.  |
> ++----------+  +---------+                   +---------+  +----------+
> +|    TSN   |  |TSN| |Svc|<-- DetNet flow -->: Service :->|    TSN   |
> ++----------+  +---+ +---+    +----------+   +---------+  +----------+
>
> Fix the right side Edge node. Service --> Svc + TSN. Same as on the left side.
>
> -- 
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/detnet-wg/draft-ietf-detnet-dp-sol-mpls/pull/24#pullrequestreview-210136145

louberger commented:
Okay. I'll clarify.


----------
On March 4, 2019 8:27:01 AM Varga Balázs <notifications@github.com> wrote:

> vb-balazs commented on this pull request.
>
>
>
>>     </t>
>     <t>
>     Similarly, an Entropy Label Indicator/Entropy Label (ELI/EL)
> -   <xref target="RFC6790"/> may be carried below the S-Label in the label 
> stack in order to control or prevent ECMP treatment for packets in a DetNet 
> flow (the use of the same EL for all of the packets in a flow forces the 
> flow to follow the same path when otherwise ECMP is in use in a network). 
> However, as previously stated in <xref target="dn-gen-encap-solution"/>, 
> the use of ECMP in a DetNet domain is not recommended.
> -	</t>
> +   <xref target="RFC6790"/> MAY be carried below the S-Label in the
> +   label stack in networks where DetNet flows would otherwise received 
> ECMP treatment.
> +   When ELs are used, the same EL value SHOULD be used for
> +   all of the packets sent using a specific S-Label to force the flow to 
> follow the
> +   same path.   However, as previously stated in <xref 
> target="dn-gen-encap-solution"/>, the use of ECMP in a DetNet domain is NOT 
> RECOMMENDED.
> +   </t>
>
> Not recommended for DetNet flows, but may be used for non-DetNet traffic.
>
> -- 
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/detnet-wg/draft-ietf-detnet-dp-sol-mpls/pull/24#pullrequestreview-210136377

louberger commented:
I agree it can be independent. The in addition was just transition text, it 
can be dropped if it is confusing.

Lou


----------
On March 4, 2019 8:27:19 AM Varga Balázs <notifications@github.com> wrote:

> vb-balazs commented on this pull request.
>
>
>
>> +       the sequence number field length may be 16 or 28 bits, and the
> +       field value can wrap.
> +     </t>
> +     <t>
> +       Note that an implementation MAY wish to constrain the maximum
> +       number sequence numbers that are tracked, on platform-wide or per
> +       flow basis.  Some implementations MAY support the provisioning of
> +       this number on either a platform-wide or per flow basis.
> +     </t>
> +   </section>
> +   <section anchor="pof-requirements" title="Packet Ordering Function
> +                                             Processing">
> +     <t>
> +       A function that is related to PEF is the Packet Ordering Function
> +       (POF).  In addition to PEF, implementations MAY support POF. When
> +       supported, use of the POF for a particular app flow MUST be
>
> POF can be also a stand alone function. This text binds it with PEF ("In 
> addition to PEF ...").
>
> -- 
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/detnet-wg/draft-ietf-detnet-dp-sol-mpls/pull/24#pullrequestreview-210136482

louberger commented:
I think this is a subtlety that can be left to the new l2 over mpls document


----------
On March 4, 2019 8:26:27 AM Varga Balázs <notifications@github.com> wrote:

> vb-balazs commented on this pull request.
>
>
>
>>  ]]></artwork>
>  </figure>
> - <t>
> -  <xref target="fig_ip_pw_detnet"/> illustrates how end to end DetNet
> -  service is provided where the end systems are able to send and receive
> -  IP DetNet flows, e.g., per <xref target="I-D.ietf-detnet-dp-sol-ip"/>, and
> -  the MPLS  nodes optionally provide service protection. In this
> -  case R1 and R3 are T-PEs and R2 is an S-PE and the DetNet service is
> -  end-to-end.
> - </t>
> +  </section>
> +  <section title="IEEE 802.1 TSN Over MPLS DetNet Data Plane Scenario"
>
> This is a more general case. It can be any L2 not only TSN. For example if 
> the IP End-systems are in the same subnet and connected via the MPLS 
> domain. Worth to add a note?
>
> -- 
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/detnet-wg/draft-ietf-detnet-dp-sol-mpls/pull/24#pullrequestreview-210136119

louberger commented:
Okay


----------
On March 4, 2019 8:26:19 AM Varga Balázs <notifications@github.com> wrote:

> vb-balazs commented on this pull request.
>
>
>
>>  
> -<figure align="center" anchor="fig_pw_detnet2"
> - title="MPLS-Based DetNet (non-MPLS End System)">
> +    <t>
> +      [Authors note: TBD - should the rest of this section be cut?]
>
> No, these are good and straight-fort  description for scenarios where "MPLS 
> to the desktop" is not possible.
>
> -- 
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/detnet-wg/draft-ietf-detnet-dp-sol-mpls/pull/24#pullrequestreview-210136047

louberger commented:
Simply that how outflows are mapped to mpls are out of scope of the 
document. I'm working on the DetNet IP over mpls section for the IP 
solution document. I should have something out late today on that.


----------
On March 4, 2019 8:26:05 AM Varga Balázs <notifications@github.com> wrote:

> vb-balazs commented on this pull request.
>
>
>
>>     function and operation of the Packet Replication (PRF) Packet Elimination
>     (PEF) and Packet Ordering (POF) functions with an MPLS data plane.
>  
>      It also describes an MPLS-based DetNet forwarding sub-layer that eliminates (or
>      reduces) contention loss  and provides bounded latency for DetNet flows.
>    </t>
>    <t>
> -   This document does not define, but does allow for, associated 
> controller plane
> -   and Operations, Administration, and Maintenance (OAM) functions.
> +    MPLS encapsulated DetNet flows can be carried over
> +    network technologies that can provide the DetNet required level
> +    of service.  This document defines examples of such, specifically
> +    carrying MPLS DetNet flows over IEEE 802.1 TSN sub-networks, and
> +    over DetNet IP PSN.
> +  </t>
> +  <t>
> +    The intent is for this document to support different traffic types
> +    being mapped over DetNet MPLS, but this is out side the scope of this
> +    document. An example of such can be found in <xref
>
> What do we ntend to say here? Does it refer to L2/L3 App-flow? Or somewhat 
> else? What is covered in the IP data-plane draft?
>
> -- 
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/detnet-wg/draft-ietf-detnet-dp-sol-mpls/pull/24#pullrequestreview-210135963

vb-balazs commented on draft-ietf-detnet-dp-sol-mpls-02.xml:
```diff
@@ -1113,37 +1004,337 @@ Examples:
+     other pseudowire encapsulation.  This implementation choice also
+     impacts PEF and POF processing as described in the next section.
    </t>
+   <section anchor="pef-requirements" title="Packet Elimination Function
```
Why no PRF section? We should highlight that in case of replication the S-label swap can result in different S-label values (i.e., member flows may have different S-label).

vb-balazs commented on draft-ietf-detnet-dp-sol-mpls-02.xml:
```diff
@@ -1113,37 +1004,337 @@ Examples:
+           </t>
+           <t>
+             MUST ensure a QoS flow exceeds its allocated resources or
+             provisioned service level,
```
I think we intend to say: "does not exceed"

vb-balazs commented on draft-ietf-detnet-dp-sol-mpls-02.xml:
```diff
@@ -2059,103 +2211,86 @@ Note: * no service sub-layer for transit nodes
+  <t>
+    To support outgoing DetNet MPLS over IP, an implementation MUST support the
+    provisioning of IP/UDP header information in place of sets of
+    F-Labels.  Note that multiple sets of F-Labels can be provisioned to
```
And also "S-labels" may differ for member flows.

vb-balazs commented on text-to-be-moved:
```diff
@@ -0,0 +1,266 @@
+      
```
Where to move this? New draft?

louberger commented:
It's included in the f-label section, we can make it a subsection if you want.


----------
On March 4, 2019 3:03:57 PM Varga Balázs <notifications@github.com> wrote:

> vb-balazs commented on this pull request.
>
>
>
>>     </t>
> +   <section anchor="pef-requirements" title="Packet Elimination Function
>
> Why no PRF section? We should highlight that in case of replication the 
> S-label swap can result in different S-label values (i.e., member flows may 
> have different S-label).
>
> -- 
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/detnet-wg/draft-ietf-detnet-dp-sol-mpls/pull/24#pullrequestreview-210340121

louberger commented:
Yes - to the IP solution document (hope to have that tonight) and a new 
l2/tsn over DetNet mpls draft...

Lou


----------
On March 4, 2019 3:24:46 PM Varga Balázs <notifications@github.com> wrote:

> vb-balazs commented on this pull request.
>
>
>
>> @@ -0,0 +1,266 @@
> +
>
> Where to move this? New draft?
>
> -- 
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/detnet-wg/draft-ietf-detnet-dp-sol-mpls/pull/24#pullrequestreview-210348692

louberger commented:
Yes.


----------
On March 4, 2019 3:15:21 PM Varga Balázs <notifications@github.com> wrote:

> vb-balazs commented on this pull request.
>
>
>
>> +         reserved DetNet flows by using resources allocated to the 
>> reserved flows.  Therefore,
> +         the network nodes of a DetNet network:
> +         <list style="symbols">
> +           <t>
> +             MUST defend the DetNet QoS by discarding or remarking (to an
> +             allocated DetNet flow or non-competing non-DetNet flow)
> +             packets received that are not the subject of a completed 
> resource allocation.
> +           </t>
> +           <t>
> +             MUST NOT use a DetNet allocated resource, e.g. a queue or 
> shaper reserved for
> +             DetNet flows, for any packet that does match the corresponding
> +             DetNet flow.
> +           </t>
> +           <t>
> +             MUST ensure a QoS flow exceeds its allocated resources or
> +             provisioned service level,
>
> I think we intend to say: "does not exceed"
>
> -- 
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/detnet-wg/draft-ietf-detnet-dp-sol-mpls/pull/24#pullrequestreview-210344835

louberger commented:
no - there is explicit text on this in the subsequent section.

On 3/4/2019 8:26 AM, Varga Balázs wrote:
>
> *@vb-balazs* commented on this pull request.
>
> ------------------------------------------------------------------------
>
> In draft-ietf-detnet-dp-sol-mpls-02.xml 
> <https://github.com/detnet-wg/draft-ietf-detnet-dp-sol-mpls/pull/24#discussion_r262056113>:
>
> >     </t>
> +   <section anchor="pef-requirements" title="Packet Elimination Function
> +                                             Processing">
> +     <t>
> +       Implementations MAY support the Packet Elimination Function (PEF)
> +       for received DetNet MPLS flows.  When supported, use of the PEF
> +       for a particular app flow MUST be provisioned via configuration
> +       and the controller plane, see <xref target="cp_considerations"/>.
> +     </t>
> +     <t>
> +       After an app flow is identified for a received DetNet MPLS
>
> Does it mean "elimination first"? Does this text set the order of 
> PREOF functions?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub 
> <https://github.com/detnet-wg/draft-ietf-detnet-dp-sol-mpls/pull/24#pullrequestreview-210136421>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/AGRWyOmCXzhLS21yPa1I4UAXIiSJhe03ks5vTR8OgaJpZM4bWAPa>.
>

louberger committed:
Address comments from Balázs
  OPEN Questions:
  - need for S-Labels to be from platform label space (Andy)
  - Different S-Labels for same application flow (Balazs)

louberger committed:
minor fix to figure

louberger committed:
PRF can't be used with 0 bit length sequence numbers

louberger committed:
minor fix

louberger committed:
changes per March 25 meeting

louberger committed:
s/MPLS DetNet/DetNet MPLS
